### ISSUE TITLE
Crash in the FFN destructor after data::Save is called

### ISSUE THREAD

**Issue by yurivict** [mlpack/mlpack#3449]
#### Issue description

This testcase crashes:
```
#include <mlpack.hpp>


int main() {
  using namespace mlpack;

  // Load the training set and testing set.
  arma::mat trainData;
  data::Load("thyroid_train.csv", trainData, true);
  arma::mat testData;
  data::Load("thyroid_test.csv", testData, true);

  arma::mat trainLabels = trainData.row(trainData.n_rows - 1) - 1;
  arma::mat testLabels = testData.row(testData.n_rows - 1) - 1;
  trainData.shed_row(trainData.n_rows - 1);
  testData.shed_row(testData.n_rows - 1);

  // Initialize the network.
  FFN<> model;
  model.Add<Linear>(10);
  model.Add<Sigmoid>();
  model.Add<Linear>(9);
  model.Add<LogSoftMax>();

  model.Train(trainData, trainLabels);
  data::Save("model.xml", "model", model, false);

  return 0;
}
```

```
(gdb) r
Starting program: /usr/home/yuri/nn-trainer/mlpack/a.out

Program received signal SIGBUS, Bus error.
Object-specific hardware error.
0x000000000024bd2b in mlpack::MultiLayer<arma::Mat<double> >::~MultiLayer (this=0x7fffffffdb80) at /usr/local/include/mlpack/methods/ann/layer/multi_layer.hpp:55
55            delete network[i];
(gdb) 
```

It doesn't crash when the data::Save call is removed.

#### Your environment

 * version of mlpack: 4.0.1
 * operating system: FfreeBSD 13.1
 * compiler: clang-14
 * version of dependencies (Boost/Armadillo): boost-libs-1.81.0, armadillo-12.0.1
 * any other environment information you think is relevant: n/a

#### Steps to reproduce
```
c++ test.cpp -g `pkg-config mlpack --cflags --libs`
```

**Comment by yurivict** [comment]
Somehow MultiLayer objects get destroyed inside of the data::Save() function call, but their pointers stay in the vector ```model.network.network```.

**Comment by aadi-raj** [comment]
Try including `#define MLPACK_ENABLE_ANN_SERIALIZATION` before including `mlpack.hpp`.

**Comment by yurivict** [comment]
This helps, thanks!

Bit it's would be better if compilation would fail so that wrong code wouldn't be generated.

**Comment by yurivict** [comment]
Thank you.

**Comment by conradsnicta** [comment]
Fixed in #3451
